### PR TITLE
Introduce Associated Labels to centroids.

### PR DESCRIPTION
**Pull request by perlfly**
This could help showing how to use KMeans as a classifier.

Some changes to KMeans properties and function.
Updated Tests accordingly.
Demo the usage of labels in one small test.

### CODE REVIEW

**Comment by hollance** [comment]
Thanks! I'm going to leave this here for the minute. I still want to look over the k-means explanation and code. Yours looks like a good improvement :-) but it might be a few days before I get around to it.

**Comment by perlfly** [comment]
Thank you @hollance . Consider it as a humble proposal. Here I add `Label` as a convenience way to show the results. K-Means is often used in machine learning as an unsupervised learning algorithm. You _train_ the centroids using some training points. Then you use the trained classifier in order to classify previously _unseen_ points. This is why I use `train` and `fit` in function names.

I see the code has almost no comment. Thus I did the same, since I don't know if this is functional to the project intentions. Maybe explanations are meant to go in the readme.

**Comment by hollance** [comment]
This repo is all about explaining stuff. Some source code in here doesn't use a lot of comments because it's not that hard to follow what's going on. But in general I like to err on the side of too many comments rather than too few.

I have a note on my to-do list that the README for k-means could do with more depth and an explanation of how the code actually works. If you're up for that, it would be a very welcome contribution. :-)

**Comment by hollance** [comment]
@perlfly: All right, I finally had some time to look at this. Merged and looked at. :-) Thanks again!

**Comment by perlfly** [comment]
@hollance  My pleasure! I hope I'll be able to do some more.
